### PR TITLE
[B+C] API for new mob spawner features. Adds BUKKIT-1998

### PR DESCRIPTION
**Pull request by betterphp**
##### The Issue:

There is no way for plugins to make use of the new spawner customisation that was introduced.
##### Justification for this PR:

Plugins may want to create custom spawners or modify the parameters of existing ones. I personally want to use this for dungeon generation.
##### PR Breakdown:
- Adds new methods to the CreatureSpawner interface to get and set the various fields.
- Adds support for all entity types that can be created using spawners to CreatureSpawner.setSpawnedType(EntityType entityType);
- Adds a method to set the properties of the spawned entity to those of an existing one, CreatureSpawner.setSpawnedEntity(Entity entity, boolean includePosition);
##### Testing Results and Materials:

Source - https://github.com/betterphp/SpawnerTest
Binary - http://ubuntuone.com/5bMZ7n3xCLxuw84on8sHDV

Provides 4 commands:
/gsp - Lists the current spawner properties
/ssp - Sets the spawner properties
/tnt - Makes a spawner than shoots TNT into the air
/fw - Makes a spawner that shoots fireworks into the air

All work on the spawner in the line of sight. A number of variations of spawner were created and their properties and behaviour checked before and after a server restart.
##### Relevant PR(s):

CB-1133 - https://github.com/Bukkit/CraftBukkit/pull/1133
##### JIRA Ticket:

BUKKIT-1998 - https://bukkit.atlassian.net/browse/BUKKIT-1998

### CODE REVIEW

**Comment by Black-Hole** [comment]
Looks nice. But I think separating the new methods to a new interface would fit better. So it could be used for SpawnerMinecart, too.

**Comment by betterphp** [comment]
@Black-Hole Good point. I'm not sure how that should be packaged though since one is a spawner and one is a block, where should the new interface go ? :/

**Comment by riking** [comment]
You changed the UNIX permissions on this file to be executable.
`100644 → 100755`

Somehow.

**Comment by riking** [comment]
@betterphp "Where should it go?" You abstract it out to another file.

A possible name would be SpawnerEntry, but that's kind of bad.

**Comment by betterphp** [comment]
@riking That was my network drive's fault :P I meant what package should it go in more than the name, org.bukkit.craftbukkit.entitiy.SpawnerEntry seems wrong because something in obc.block would implements it :/

**Comment by riking** [comment]
I think it needs to go in `org.bukkit`.

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 1]
```diff
@@ -85,4 +101,88 @@
...
+
+    /**
+     * Sets the number of other entities that need to be within the spawner's range to prevent spawning.
+     * 
+     * @param maxNearbyEntities The number.
+     */
+    public void setMaxNearbyEntities(int maxNearbyEntities);
+
+    /**
+     * Gets the range that a player must be within to allow spawning.
```
Comment (riking) [inline src/main/java/org/bukkit/block/CreatureSpawner.java]: Document that this is a sphere.

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 2]
```diff
@@ -85,4 +101,88 @@
...
+
+    /**
+     * Sets the number of entities that the spawner will attempt to spawn.
+     * 
+     * @param count The number.
+     */
+    public void setCount(int count);
+
+    /**
+     * Gets the maximum distance from the spawner that entities will be spawned.
```
Comment (riking) [inline src/main/java/org/bukkit/block/CreatureSpawner.java]: Document that this is a horizontal square.

Comment (riking) [inline src/main/java/org/bukkit/block/CreatureSpawner.java]: Actually, it's a square triangular distribution.

**Comment by riking** [comment]
Okay, so the API for an entry in the entity list. It is an analogue for [TileEntityMobSpawnerData](https://github.com/Bukkit/mc-dev/blob/master/net/minecraft/server/TileEntityMobSpawnerData.java).

`org.bukkit.SpawnerEntry` (Working name)

``` java
public class SpawnerEntry {
    public SpawnerEntry(int weight, 
}
```

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 3]
```diff
@@ -28,26 +41,73 @@
...
-     * Set the spawner creature type.
+     * Sets the entity that will be spawned.
      *
-     * @param creatureType The creature type.
-     * @deprecated In favour of {@link #setSpawnedType(EntityType)}.
+     * @param entity The entity
+     * @throws IllegalArgumentException If the entity cannot be spawned from spawners.
      */
-    @Deprecated
-    public void setCreatureType(CreatureType creatureType);
```
Comment (riking) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:41]: Why'd you move this function?

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 4]
```diff
@@ -28,26 +41,73 @@
...
+     * Sets the entity that will be spawned.
      *
-     * @return The creature type's name.
-     * @deprecated Use {@link #getCreatureTypeName()}.
+     * @param entity The entity.
+     * @param includePosition If true position data will be kept.
+     * @throws IllegalArgumentException If the entity cannot be spawned from spawners.
      */
-    @Deprecated
-    public String getCreatureTypeId();
```
Comment (riking) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:50]: You moved this one as well

File: src/main/java/org/bukkit/block/CreatureSpawner.java (hunk 3 above)
Comment (betterphp) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:41]: The get and set had become separated by some other methods. It made sense to keep them together.

File: src/main/java/org/bukkit/block/CreatureSpawner.java (hunk 4 above)
Comment (betterphp) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:50]: same as above.

File: src/main/java/org/bukkit/block/CreatureSpawner.java (hunk 3 above)
Comment (riking) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:41]: We want to keep diffs to a minimum.

Comment (EdGruberman) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:41]: @riking mostly this concern applies to nms.  Organizing other code outside of what is affected by a version update effort is fine.  This type of situation would be fine to resort/reorder things if it makes sense.

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 5]
```diff
@@ -85,4 +154,103 @@
...
+     * in the range specified. Vanilla behavior (1 below, 1 above) is equivalent to calling this with (1, 1).
+     *
+     * @param maxBelow Maximum Y-difference below this CreatureSpawner for spawning attempts
+     * @param maxAbove Maximum Y-difference above this CreatureSpawner for spawning attempts
+     * @throws IllegalArgumentException when no Y-values would be possible (e.g. setVerticalRange(1, -2))
+     */
+    public void setVerticalRange(int maxBelow, int maxAbove);
+
+    /**
+     * Gets the number of other entities that need to be within the spawner's range to prevent spawning,
```
Comment (ST-DDT) [inline src/main/java/org/bukkit/block/CreatureSpawner.java]: Minor typo at the end of this line `,` => `.`

Comment (betterphp) [inline src/main/java/org/bukkit/block/CreatureSpawner.java]: Oops, good spot :P

**Comment by riking** [comment]
I think plugins deserve a way to construct a SpawnerEntry from other information than just the entity type.. like PotionEffects, etc.... but I can't think of a good API way to do it.

**Comment by betterphp** [comment]
I agree, the only way I could think of was something like ItemMeta but for entities. It would basically be a wrapper for the NBT tags for each mob.

So plugins could do somethign along the lines of.

``` java
CreatureSpawner spawner = (CreatureSpawner) block.getState();
CreeperMeta creeper = EntityType.CREEPER.createMeta();

creeper.setLocation(spawner.getLocation());
creeper.setPowered(true);

spawner.addSpawnEntry(new SpawnerEntry(creeper));
```

I guess you would then add methods to Entity to apply it to them in the same way ItemStacks have for ItemMeta.

**Comment by riking** [comment]
@betterphp We would want something that looks more like createNewSpawnerEntry(creeper) that could be overridden in CB for the last line

**Comment by betterphp** [comment]
Didn't you ask for it to be how it is in my example here https://github.com/Bukkit/Bukkit/pull/849#issuecomment-16136613 or did I misunderstand something ? The creeper variable above isn't an entity.

**Comment by riking** [comment]
OOOOhhhh missed that

**Comment by ST-DDT** [comment]
How about calling it SpawnTask (interface) with only a single method
public void spawn(Location location);
or
public void spawn(CreatureSpawner spawner);

and CreatureSpawner.setSpawnTask(SpawnTask task);

no other methods at all. Everything else could be handled inside the spawn tasks. (Weight, Spawned Types/Creatures etc)

In addition to that there should be a BukkitSpawnTask which would contain basic/default (vanilla like) spawn behaviour. 
This would be similar to MetaDataValue and FixedMetaDataValue.

public void setSpawnedType(EntityType creatureType)
{
setSpawnTask(new BukkitSpawnTask(creatureType));
or
setSpawnTask(SpawnTaskFactory.getByCreature(creatureType));
}

in CraftBukkit you will just call task.spawn(this);

**Comment by betterphp** [comment]
:P okay. Well what do you think of the meta idea overall ? I can see how it would work but it's a bit of a big thing to do if it's just going to get rejected.

@ST-DDT I don't really see the need to add that extra complication, it doesn't help with the entity problem. It would allow plugins to create one spawner config and apply it to multiple blocks, but then so would a method that they add that sets up the spawner.

**Comment by ST-DDT** [comment]
No, the SpawnTask won't configure the spawner.
It would just spawn the entities.
You would be able to spawn creatures with custom health, equip and
everything.

It should implement something like Serializable or
ConfigurationSerializable to save the task to disk and load it from
disk, (similar to ItemStack/ItemMeta) but nothing else would be required.
Using the way suggested in this PR would only result in limitations
which cannot be reverted by future updates.

I implemented something similar for one of my plugins. CrazySpawner.
There are almost one hundred possebilites to spawn the default
creatures. ColoredSheep, ChargedCreepers, AngryWolves, Butcher
(Villager), Spider_Jokeys etc and i didn't use any special abilites for
that.

**Comment by betterphp** [comment]
Oh I see, I still prefer modifying the spawner instead plugins providing their own method. You can get all of the combinations you suggested using this method without the need for any saving/loading of extra files. The spawner copies the stored NBT tag onto the created entity so there should not be any limitations at all really :/ Plus if you are having it done in code, why not use the SpawnerSpawnEvent that was proposed in another PR ?

**Comment by ST-DDT** [comment]
I guess your Meta part is (almost) similar to my idea.

CreeperMeta creeper = EntityType.CREEPER.createMeta();
You will have to type cast here. But it would be also possible to use new CreeperMeta() directly due to this being completly inside Bukkit without any MC or CB references. (if the constructor isn't hidden somewhy)
(CreeperMeta implements/extends CreatureMeta, in my previous post i just called CreatureMeta SpawnTask (combined with your SpawnEntry)).

creeper.setLocation(spawner.getLocation());
what's this method for?
forcing the creeper to spawn at a specific location?

**Comment by betterphp** [comment]
An EntityMeta is a wrapper for the NBT data so it would have to be created in CraftBukkit, so CraftCreeperMeta would hold an NBTCompound and have a method to get that so it can be used in the spawners list. The Bukkit side would be an interface to set the various tags that each mob has so the setLocation() method would set the Pos tag. Basically it exposes all of the properties that entities can have without having to have an actual entity in the world.

**Comment by ST-DDT** [comment]
Okay, but then EntityType.CREEPER.createMeta(); won't work that simple.
You will have to redirect that call to a factory.
Either 
public EntityMeta createMeta()
{
return Bukkit.getServer().createEntityMeta(this);
or
return Bukkit.getEntityMetaFactory().createEntityMeta(this);
or 
return METAFACTORY.createEntityMeta(this);
//the last two require a static field for the factory instance which has to be set via CB (similar to the Enchantments being initialized and locked during CB startup)
}

**Comment by betterphp** [comment]
I had the getEntityMetaFactory() method in mind. I haven't looked into it properly but I believe that is the same way ItemMeta works ?

**Comment by ST-DDT** [comment]
(Craft)ItemMeta are created inside CraftItemStack, but there is no such CB thing in the enum class.
EDIT start
I have been wrong.
Bukkit.getItemFactory().asMetaFor(meta, getType0()); (ItemStack: 181) 
EDIT end

In my opinion the meta should be created this way:
return Bukkit.getServer().createEntityMeta(this);

this would allow using a static methods without creating an instance.

CraftServer:
public EntityMeta createEntityMeta(EntityType type)
{
return CraftEntityMetaFactory.createEntityMeta(type);
}

EDIT: If we stick to Bukkits previous behaviour, your choice will be the right one. (getEntityFactory())

**Comment by EdGruberman** [comment]
SpawnDefinition would be a good name.  Entry would seem to imply multiple could exist per spawner.  Can they?

Not sure I see the use case for a SpawnDefinition class to implement ConfigurationSerializable at all.

I also sorta hate the term "meta".  It's non-descriptive and overly abused.  I would think CreatureSpawner.createDefinition would be a great static factory method.  I don't see why you'd force it into Bukkit or Server.

I can't guarantee core team would approve though... just my random two cents in a drive by on this conversation.

**Comment by betterphp** [comment]
@EdGruberman You can add more than one entry to a single spawner with a weight for each one. If we are open to naming discussions though I liked SpawnPotential since this object represents one of the potential spawns :P

I see what you mean about the meta thing, but don't really agree with it being CreatureSpawner.createDefinition() since that would limit it to spawners only making Entity.setMeta(EntityMeta) seem a bit wrong even though it would still be a valid and useful thing to have availab

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 6]
```diff
@@ -85,4 +146,103 @@
...
+    /**
+     * Sets the maximum distance from the spawner that entities will be spawned. This uses a horizontal
+     * square centred on the spawner
+     *
+     * @param range The distance.
+     */
+    public void setRange(int range);
+
+    /**
+     * Set the vertical range of this CreatureSpawner. Spawn attempts will be uniformly distributed
```
Comment (natemort) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:210]: Should be "Sets".

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 7]
```diff
@@ -85,4 +146,103 @@
      * @param delay The delay.
      */
     public void setDelay(int delay);
+
+    /**
+     * Get the spawner's minimum delay.
```
Comment (natemort) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:151]: Should be "Gets".

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 8]
```diff
@@ -85,4 +146,103 @@
...
+     *
+     * @return The delay.
+     */
+    public int getMinDelay();
+
+    /**
+     * Set the spawner's minimum delay.
```
Comment (natemort) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:158]: Should be "Sets".

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 9]
```diff
@@ -85,4 +146,103 @@
...
+     *
+     * @param delay The delay.
+     * @throws IllegalArgumentException if delay is less than or equal to 0
+     */
+    public void setMinDelay(int delay);
+
+    /**
+     * Get the spawner's maximum delay.
```
Comment (natemort) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:166]: Should be "Gets".

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 10]
```diff
@@ -85,4 +146,103 @@
...
+     *
+     * @return The delay.
+     */
+    public int getMaxDelay();
+
+    /**
+     * Set the spawner's maximum delay.
```
Comment (natemort) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:173]: Should be "Sets".

**Comment by riking** [comment]
What if we had something like a AnimalSpawnerEntry with the fields contained in Animal (so age, is adult, etc)?
And a ProjectileSpawnerEntry that remembers the Shooter?

What if we had World.spawnEntity(SpawnerEntry) which just executes one of these?

File: src/main/java/org/bukkit/block/CreatureSpawner.java [hunk 11]
```diff
@@ -28,26 +41,65 @@
...
+    /**
+     * Creates a new {@link SpawnerEntry} from an entity, when added to a spawner all properties of the
+     * entity will be included. The position data can be removed by setting the position flag to false.
+     *
+     * @param entity The entity
+     * @param includePosition The position flag
+     * @param weight The weight
+     * @return The entry
+     */
+    public SpawnerEntry createSpawnerEntry(Entity entity, boolean includePosition, int weight);
```
Comment (natemort) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:87]: I'm somewhat hesitant about SpawnerEntries that can be created with Entity data that can't later be accessed.  At the same time, however, creating a fully functional Entity data system is a much larger addition, and without a doubt outside of the scope of this change.

**Comment by betterphp** [comment]
@riking That's still not sufficient, people will want to take advantage of all of the properties or each entity not just the general ones. Also why restrict it to spawning, being able to apply the data to existing mobs would be useful too.

File: src/main/java/org/bukkit/block/CreatureSpawner.java (hunk 11 above)
Comment (betterphp) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:87]: I agree, not really sure what you're suggesting there though ?

File: src/main/java/org/bukkit/block/CreatureSpawner.java (hunk 10 above)
Comment (betterphp) [inline src/main/java/org/bukkit/block/CreatureSpawner.java:173]: Sorry but it has to be said, could these very minor issues really not have been mentioned 4 months ago ? Will make the changes if the other issues can be resolved.

**Comment by betterphp** [comment]
Enough time has been wasted on this.

**Pull request closed without merging**